devices/AtomQrCode/QrCodeReader.cs

```diff
@@ -0,0 +1,843 @@
+        }
+
+        /// <summary>
+        /// Indicates that barcode data is available from the <see cref="QrCodeReader"/> object.
+        /// </summary>
+        public event BarcodeDataAvailableEventHandler BarcodeDataAvailable
+        {
+            add
+            {
+                if (_disposed)
+                {
+#pragma warning disable S3877 // OK to throw this here
+                    throw new ObjectDisposedException();
+#pragma warning restore S3877 // Exceptions should not be thrown from unexpected methods
+                }
+
+                BarcodeDataAvailableEventHandler callbacksOld = _callbacksBarcodeDataAvailableEvent;
+                BarcodeDataAvailableEventHandler callbacksNew = (BarcodeDataAvailableEventHandler)Delegate.Combine(callbacksOld, value);
+
+                try
+                {
+                    _callbacksBarcodeDataAvailableEvent = callbacksNew;
+                }
+                catch
+                {
+                    _callbacksBarcodeDataAvailableEvent = callbacksOld;
+
+                    throw;
+                }
+            }
+
+            remove
+            {
+                if (_disposed)
+                {
+#pragma warning disable S3877 // OK to throw this here
+                    throw new ObjectDisposedException();
+#pragma warning restore S3877 // Exceptions should not be thrown from unexpected methods
+                }
+
+                BarcodeDataAvailableEventHandler callbacksOld = _callbacksBarcodeDataAvailableEvent;
+                BarcodeDataAvailableEventHandler callbacksNew = (BarcodeDataAvailableEventHandler)Delegate.Remove(callbacksOld, value);
+
+                try
+                {
+                    _callbacksBarcodeDataAvailableEvent = callbacksNew;
+                }
+                catch
+                {
+                    _callbacksBarcodeDataAvailableEvent = callbacksOld;
+
+                    throw;
+                }
+            }
+        }
+
+
+        /// <summary>
+        /// Handles internal events and re-dispatches them to the publicly subscribed delegates.
+        /// </summary>
+        /// <param name="barcodeData">The barcode data.</param>
+        internal void OnBarcodeDataAvailableInternal(string barcodeData)
+        {
+            BarcodeDataAvailableEventHandler callbacks = null;
+
+            if (!_disposed && !string.IsNullOrEmpty(barcodeData))
+            {
+                callbacks = _callbacksBarcodeDataAvailableEvent;
+            }
+
+            callbacks?.Invoke(this, new BarcodeDataAvailableEventArgs(barcodeData));
+        }
+
+#region Dispose implementation
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    if (_readerSerialPort != null)
+                    {
+                        _readerSerialPort.Close();
```

Comment:
_readerSerialPort?.Close();
same for the newt one and get rid of the if null? (just for simplification)